Rust/src/common/cluster.rs

```diff
@@ -0,0 +1,338 @@
+use std::cmp::{max, min};
+use rand::{Rng, SeedableRng};
+use rand_chacha::ChaCha20Rng;
+use rayon::iter::{IntoParallelRefMutIterator, ParallelIterator};
+use crate::common::samplesummary::SampleSummary;
+
+const PHASE2_THRESHOLD : usize = 2;
+const separation_ratio_for_merge : f64 = 0.8;
+const weight_threshold : f64 = 1.25;
+const LENGTH_BOUND : usize = 1000;
+
+pub trait Cluster<Q,T: ?Sized> {
```

Comment:
We should document the trait functions. If someone wants to create a new cluster implementation, they need to know how they're expected to implement `recompute`, `reset`, etc.

Reply:
Agree. Done.